### PR TITLE
Improve support for IAR compiler with attributes and intrinsics

### PR DESCRIPTION
**Pull request by josepho0918**
This patch improves development experience and performance for projects using the IAR toolchain.

- Utilize IAR compiler attributes
- Optimized with IAR intrinsics

### CODE REVIEW

File: lib/common/entropy_common.c [hunk 1]
```diff
@@ -38,7 +38,11 @@ const char* HUF_getErrorName(size_t code) { return ERR_getErrorName(code); }
 /*-**************************************************************
 *  FSE NCount encoding-decoding
 ****************************************************************/
+#if defined(__IAR_SYSTEMS_ICC__)
+MEM_STATIC
```
Comment (Cyan4973) [inline lib/common/entropy_common.c]: Is there a problem with `FORCE_INLINE_TEMPLATE` ?
Why is it a problem here specifically ?

**Comment by Cyan4973** [comment]
Sometimes I see `__ICCARM__` being tested, and sometimes `__IAR_SYSTEMS_ICC__`.
Would you mind explaining the difference, and why sometimes one is better than the other ?

File: lib/common/entropy_common.c (hunk 1 above)
Comment (josepho0918) [reply lib/common/entropy_common.c]: > Is there a problem with `FORCE_INLINE_TEMPLATE` ? Why is it a problem here specifically ?

The IAR compiler complains that the `FSE_readNCount_body` function is too complex to be inlined.

**Comment by josepho0918** [comment]
> Sometimes I see `__ICCARM__` being tested, and sometimes `__IAR_SYSTEMS_ICC__`. Would you mind explaining the difference, and why sometimes one is better than the other ?

`__IAR_SYSTEMS_ICC__` identifies the IAR compiler platform, while `__ICCARM__` specifically designates the IAR compiler for Arm architecture. Since there're IAR versions for various architectures, such as RISC-V, `__ICCARM__` is utilized for IAR intrinsics.

File: lib/common/entropy_common.c (hunk 1 above)
Comment (Cyan4973) [reply lib/common/entropy_common.c]: The function is indeed pretty big.

It seems the point of `FORCE_INLINE_TEMPLATE` is to pass down `BMI2_TARGET_ATTRIBUTE` compilation directive to an instance of the function, which is invoked when the appropriate target (`AVX2` + `BMI2` architecture) is detected at runtime.

Of course, this capability is irrelevant for `ARM` architectures.

So yes, technically this modification works. It's just a pity that it translates into one exception which will have to be maintained for one compiler.

**Comment by Cyan4973** [comment]
Quick question : 
does it sound plausible to feature an `__IAR_SYSTEMS_ICC__` compiler test during CI tests on Github Actions ?

If not, this is a "blind fix", and any future update could break compatibility is subtle ways, which would remain invisible.

**Comment by josepho0918** [comment]
> Quick question : does it sound plausible to feature an `__IAR_SYSTEMS_ICC__` compiler test during CI tests on Github Actions ?
> 
> If not, this is a "blind fix", and any future update could break compatibility is subtle ways, which would remain invisible.

It will be fantastic to have CI tests for IAR. However, it's a paid tool.

File: lib/common/entropy_common.c (hunk 1 above)
Comment (josepho0918) [reply lib/common/entropy_common.c]: Would it be better to use `HINT_INLINE` instead of `FORCE_INLINE_TEMPLATE` for `FSE_readNCount_body`?

**Review by Cyan4973 — APPROVED** [review]

**Pull request merged**